core/src/smartcontracts/wasm.rs

```diff
@@ -841,113 +929,323 @@ impl<'wrld> Runtime<state::Trigger<'wrld>> {
     }
 }
 
-impl DefaultExecute for Runtime<state::Trigger<'_>> {}
+impl<'wrld> import::ExecuteOperations<state::Trigger<'wrld>> for Runtime<state::Trigger<'wrld>> {
+    #[codec::wrap]
+    fn execute_query(
+        query: QueryBox,
+        state: &state::Trigger<'wrld>,
+    ) -> Result<Value, ValidationFail> {
+        iroha_logger::debug!(%query, "Executing as trigger");
+
+        let wsv: &WorldStateView = state.common.wsv;
 
-impl<'wrld> Runtime<state::validator::Validate<'wrld>> {
-    /// Execute the given module of runtime validator
+        // NOTE: Smart contract (not validator) is trying to execute the query, validate it first
+        // TODO: Validation should be skipped when executing smart contract.
+        // There should be two steps validation and execution. First smart contract
+        // is validated and then it's executed. Here it's validating in both steps.
+        // Add a flag indicating whether smart contract is being validated or executed
+        wsv.validator()
+            .validate_query(wsv, state.authority(), query.clone())?;
+
+        query
+            .execute(wsv)
+            .map(|lazy_value| match lazy_value {
+                LazyValue::Value(value) => value,
+                LazyValue::Iter(iter) => Value::Vec(iter.collect()),
+            })
+            .map_err(Into::into)
+    }
+
+    #[codec::wrap]
+    fn execute_instruction(
+        instruction: InstructionBox,
+        state: &mut state::Trigger<'wrld>,
+    ) -> Result<(), ValidationFail> {
+        debug!(%instruction, "Executing as Trigger");
+
+        // TODO: Validation should be skipped when executing smart contract.
+        // There should be two steps validation and execution. First smart contract
+        // is validated and then it's executed. Here it's validating in both steps.
+        // Add a flag indicating whether smart contract is being validated or executed
+        let wsv: &mut WorldStateView = state.common.wsv;
+        wsv.validator()
+                .clone() // Cloning validator is a cheap operation
+                .validate_instruction(wsv, &state.common.authority, instruction)
```

Comment:
This code fragment literally replicates the fragment on lines +888 to +895. Since the expanded comment is also replicated, suggest moving it into a separate function.